### PR TITLE
Implemented isin_Tensor_Tensor_out for MPS backend

### PR DESCRIPTION
**Pull request by jhavukainen**
Addresses issue #124518, adds isin_Tensor_Tensor_out.

Tests added to test_mps.py.

### CODE REVIEW

File: test/test_mps.py [hunk 1]
```diff
@@ -8148,6 +8147,28 @@ def helper(dtype):
 
         [helper(dtype) for dtype in [torch.float32, torch.float16, torch.int32, torch.int16, torch.uint8, torch.int8, torch.bool]]
 
+    def test_isin(self):
+        def helper(dtype):
+            shapes = [([2, 5], [3, 5, 2]), ([10, 3, 5], [20, 1, 3]), ([5], [10])]
+            for shape_tuple in shapes:
+                for inverted in [True, False]:
+                    if dtype == torch.float32 or dtype == torch.float16:
```
Comment (malfet) [inline test/test_mps.py:8155]:
```suggestion
                    if dtype.is_floating_point:
```

File: test/test_mps.py [hunk 2]
```diff
@@ -8148,6 +8147,28 @@ def helper(dtype):
...
+
+                    A_mps = A.clone().detach().to('mps')
+                    B_mps = B.clone().detach().to('mps')
+
+                    cpu_ref = torch.isin(A, B, invert=inverted)
+                    mps_out = torch.isin(A_mps, B_mps, invert=inverted)
+
+                    self.assertEqual(mps_out, cpu_ref)
+
+        [helper(dtype) for dtype in [torch.float32, torch.int32, torch.int16, torch.uint8, torch.int8]]
```
Comment (malfet) [inline test/test_mps.py:8170]: You are missing float16 here
```suggestion
        for dtype in [torch.float32, torch.float16, torch.int32, torch.int16, torch.uint8, torch.int8]:
            helper(dtype)
```

File: aten/src/ATen/native/mps/operations/TensorCompare.mm [hunk 3]
```diff
@@ -268,6 +269,52 @@ static void clamp_scalar_out_mps(const Tensor& input_t,
   }
 }
 
+static void isin_Tensor_Tensor_out_mps(const Tensor& elements,
+                                       const Tensor& test_elements,
+                                       bool assume_unique,
+                                       bool invert,
+                                       const Tensor& out,
+                                       string op_name) {
+  if (elements.numel() == 0) {
```
Comment (malfet) [inline aten/src/ATen/native/mps/operations/TensorCompare.mm:278]: Please add `TORCH_CHECK(elements.is_mps() && test_elements.is_mps())` as well as `TORCH_CHECK(elements.dtype() == test_elements.dtype())`

Comment (malfet) [inline aten/src/ATen/native/mps/operations/TensorCompare.mm:278]: Don't you want to check if `test_elements` is empty than you return True?

**Review by malfet — APPROVED** [review]
LGTM, see few comments

File: aten/src/ATen/native/mps/operations/TensorCompare.mm (hunk 3 above)
Comment (jhavukainen) [reply aten/src/ATen/native/mps/operations/TensorCompare.mm:279]: Added the checks and tests for them.

File: test/test_mps.py (hunk 2 above)
Comment (jhavukainen) [reply test/test_mps.py]: Added the test. Had to circumvent a missing FP16 implementation on CPU for isin through computing as FP32 and downcasting it to FP16 for the comparison check, but taking into account the nature of the op this shouldn't cause problems.

File: aten/src/ATen/native/mps/operations/TensorCompare.mm (hunk 3 above)
Comment (jhavukainen) [reply aten/src/ATen/native/mps/operations/TensorCompare.mm:279]: Yeah good point, the implementation was not handling empty `test_elements` correctly. Included this as well as empty `elements` in the test cases and added the correct handling.

**Review by kulinseth — APPROVED** [review]

**Comment by malfet** [comment]
While this PR waiting to be landed workaround have already been merged see https://github.com/huggingface/transformers/pull/30376

**Comment by kulinseth** [comment]
> While this PR waiting to be landed workaround have already been merged see [huggingface/transformers#30376](https://github.com/huggingface/transformers/pull/30376)

Yeah @jhavukainen  is moving those tests to xfail for MacOS 13 and make sure its working correctly on MacOS14.

File: test/test_mps.py [hunk 4]
```diff
@@ -8148,6 +8150,46 @@ def helper(dtype):
...
+                    cpu_ref = torch.isin(A, B, invert=inverted)
+                    if dtype is torch.float16:
+                        cpu_ref.type(dtype)
+
+                    mps_out = torch.isin(A_mps, B_mps, invert=inverted)
+                    self.assertEqual(mps_out, cpu_ref)
+
+        [helper(dtype) for dtype in [torch.float32, torch.float16, torch.int32, torch.int16, torch.uint8, torch.int8]]
+
+    @unittest.skipIf(product_version < 14.0, "Skipped on MacOS < 14.0")
```
Comment (malfet) [inline test/test_mps.py:8180]: We should not skip, but rather xfail

Comment (kulinseth) [reply test/test_mps.py:8180]: cc @jhavukainen

Comment (jhavukainen) [reply test/test_mps.py:8180]: https://github.com/pytorch/pytorch/pull/125412

**Comment by nandinimundra** [comment]
HI, I am facing this issue on M3 pro chip, While using Gemma-2 9B model. Please let me know how to resolve this.

Thanks, 
Nandini

**Pull request closed without merging**